File: word_pattern.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    bool wordPattern(string pattern, string s)
    {

        vector<string> strs;

        stringstream ss(s);

        while (!ss.eof())
        {
            string str;
            getline(ss, str, ' ');
            strs.push_back(str);
        }

        unordered_map<char, string> mp;

        if (strs.size() != pattern.size())
        {

            return false;
        }

        for (int i = 0; i < pattern.size(); i++)
        {

            if (mp.find(pattern[i]) == mp.end())
            {

                for (auto it = mp.begin(); it != mp.end(); it++)
                {

                    if (it->second == strs[i])
                    {

                        return false;
                    }
                }
                mp.insert({pattern[i], strs[i]});
            }
            else
            {

                auto it = mp.find(pattern.at(i));
                if (it->second != strs[i])
                {
                    return false;
                }
            }
        }
        return true;
    }
};

int main()
{

    Solution s;
    cout << s.wordPattern("abba", "dog cat cat dog") << endl;

    return 0;
}

File: Word_pattern.java

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

// problem link: https://leetcode.com/problems/word-pattern/
public class Word_pattern {

    public static void main(String[] args) {
        Solution s = new Solution();

        System.out.println(s.wordPattern("abba", "cat cat cat cat"));
    }

}

class Solution {

    public boolean wordPattern(String pattern, String s) {
        List<String> strs = new ArrayList<>(List.of(s.split(" ")));
        HashMap<Character, String> mp = new HashMap<>();
        if (pattern.length() != strs.size()) {

            return false;
        }

        for (int i = 0; i < pattern.length(); i++) {

            if (!mp.containsKey(pattern.charAt(i))) {
                var it = mp.entrySet().iterator();
                while (it.hasNext()) {

                    var item = it.next();
                    if (item.getValue().equals(strs.get(i))) {
                        return false;
                    }
                }
                mp.put(pattern.charAt(i), strs.get(i));

            } else {

                if (!mp.get(pattern.charAt(i)).equals(strs.get(i))) {
                    System.out.println(i);
                    return false;
                } else {
                    mp.put(pattern.charAt(i), strs.get(i));
                }

            }
        }

        return true;

    }

}
